File: basic_dsa_two_hundred_forty_two.cpp
/**
 * Basic DSA - Squares of a Sorted Array (LeetCode 977)
 * Return squares of each element in non-decreasing order.
 */
#include <iostream>
#include <vector>
#include <algorithm>

std::vector<int> sortedSquares(std::vector<int>& nums) {
    for (int& x : nums) x *= x;
    std::sort(nums.begin(), nums.end());
    return nums;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Squares of a Sorted Array\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {-4, -1, 0, 3, 10};
    sortedSquares(v1);
    std::cout << "[-4,-1,0,3,10] -> [";
    for (int i = 0; i < (int)v1.size(); i++) std::cout << (i ? "," : "") << v1[i];
    std::cout << "]\n";

    std::vector<int> v2 = {-7, -3, 2, 3, 11};
    sortedSquares(v2);
    std::cout << "[-7,-3,2,3,11] -> [";
    for (int i = 0; i < (int)v2.size(); i++) std::cout << (i ? "," : "") << v2[i];
    std::cout << "]\n";
}

File: basic_dsa_two_hundred_forty_two.py
"""
Basic DSA - Next Greater Element II
Given a circular integer array nums, return the next greater number for every element.
The next greater number of a number x is the first greater number to its traversing-order
next in the array (wrapping around). If it doesn't exist, return -1.
"""


def next_greater_elements(nums):
    """
    Monotonic decreasing stack of indices over 2 passes.
    Time Complexity: O(n)
    Space Complexity: O(n)
    """
    n = len(nums)
    res = [-1] * n
    stack = []
    for i in range(2 * n):
        idx = i % n
        while stack and nums[stack[-1]] < nums[idx]:
            res[stack.pop()] = nums[idx]
        if i < n:
            stack.append(idx)
    return res


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Next Greater Element II")
    print("=" * 60)

    print(next_greater_elements([1, 2, 1]))  # [2, -1, 2]
    print(next_greater_elements([1, 2, 3, 4, 3]))  # [2, 3, 4, -1, 4]
